### PR TITLE
Restrict Java codegen to types that can be seen over the Ledger API

### PR DESCRIPTION
**Pull request by stefanobaghino-da**
Fixes https://github.com/digital-asset/daml/issues/13324

Builds on top of refactorings part of https://github.com/digital-asset/daml/pull/13372, https://github.com/digital-asset/daml/pull/13439 and https://github.com/digital-asset/daml/pull/13440,
alongside adding interface support to the transitive closure function
in https://github.com/digital-asset/daml/pull/13557.

changelog_begin
[Java codegen] The Java codegen now restricts its output to Daml definitions
that are either directly or indirectly referred as part of a template or
interface (that is, only the types that can be seen over the Ledger API).
See https://github.com/digital-asset/daml/issues/13324 for further details.
changelog_end

### CODE REVIEW

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/ClassForType.scala [hunk 1]
```diff
@@ -4,9 +4,9 @@
 package com.daml.lf.codegen.backend.java.inner
 import com.daml.lf.codegen.TypeWithContext
 import com.daml.lf.codegen.backend.java.JavaEscaper
-import com.daml.lf.data.Ref.PackageId
-import com.daml.lf.iface.InterfaceType.{Normal, Template}
-import com.daml.lf.iface.{Enum, DefDataType, InterfaceType, Record, Variant}
```
Comment (stefanobaghino-da) [inline language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/ClassForType.scala:9]: Unnecessary diff noise.

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/CodeGenRunner.scala [hunk 2]
```diff
@@ -91,13 +79,12 @@ private final class CodeGenRunner(
...
-  ): Iterable[JavaFile] = {
-    MDC.put("packageId", moduleWithContext.packageId)
-    MDC.put("packageIdShort", moduleWithContext.packageId.take(7))
-    MDC.put("moduleName", moduleWithContext.name)
-    val typeSpecs = moduleWithContext.typesLineages.flatMap(ClassForType(_, packagePrefixes))
+  private def createTypeDefinitionClasses(module: ModuleWithContext): Iterable[JavaFile] = {
+    MDC.put("packageId", module.packageId)
+    MDC.put("packageIdShort", module.packageId.take(7))
+    MDC.put("moduleName", module.name)
+    val relevant = scope.serializableTypes.view.map(_._1).toSet
```
Comment (stefanobaghino-da) [inline language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/CodeGenRunner.scala:86]: This can probably be computed once and re-used.

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/ClassForType.scala [hunk 3]
```diff
@@ -15,21 +15,27 @@ object ClassForType extends StrictLogging {
...
+      val className = ClassName.bestGuess(classNameString)
+      generateInterfaceTypes(typeWithContext, packagePrefixes) ++
+        generateSerializableTypes(typeWithContext, className, packagePrefixes, lfInterfaceType)
+    }
+
+    if (relevant(typeWithContext.identifier)) {
+      typeWithContext.`type`.typ.fold(recurOnTypeLineages)(generateForType)
+    } else {
+      recurOnTypeLineages
+    }
```
Comment (stefanobaghino-da) [inline language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/ClassForType.scala:37]:
```suggestion
    Option
      .when(relevant(typeWithContext.identifier))(typeWithContext.`type`.typ)
      .flatten
      .fold(recurOnTypeLineages)(generateForType)
```
Avoid invoking the recursion from two different places.

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/CodeGenRunner.scala (hunk 2 above)
Comment (stefanobaghino-da) [reply language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/CodeGenRunner.scala]: Addressed by https://github.com/digital-asset/daml/pull/13571/commits/9fb5d981e35f32e6a284d2ce93c0d8f203273121

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/ClassForType.scala (hunk 1 above)
Comment (stefanobaghino-da) [reply language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/ClassForType.scala]: Addressed by https://github.com/digital-asset/daml/pull/13571/commits/30c177b6e86d741683878ab7ee0efdaebd4f5749

**Review by S11001001 — APPROVED** [review]
Nice punctuation for the end of a long sequence of PRs. 👍
